### PR TITLE
Enable support for istitle() method for unicode string

### PR DESCRIPTION
**Pull request by Vyacheslav-Smirnov**
The changes are done upon the https://github.com/numba/numba/pull/4442 PR
 - Refactor code and algorithm
 - Update documentation

### CODE REVIEW

**Comment by Vyacheslav-Smirnov** [comment]
Hi, please set the 'In progress' label for this PR

**Comment by Vyacheslav-Smirnov** [comment]
@stuartarchibald , Regarding your comment https://github.com/numba/numba/pull/4442#discussion_r314353276 , I do not think we can get good profit from generation of those isFoo algorithms: They are small enough and have only loop on string that can be extracted. Think that it would be easier and clear to implement them directly.
@shssf , What do you think?

**Comment by Vyacheslav-Smirnov** [comment]
@shssf , @densmirn , could you please review the changes?

File: numba/tests/test_unicode.py [hunk 1]
```diff
@@ -904,6 +908,40 @@ def try_compile_bad_optional(*args):
...
+            '%^Abc 5 $% Def'
+            '𐐁𐐩',
+            '𐐧𐑎',
+            '𐐩',
+            '𐑎',
+            '🐍 Is',
+            '🐍 NOT',
+            '👯Is'
+        ]
+        ISTITLE_EXAMPLES = UNICODE_EXAMPLES + unicode_title + special
```
Comment (densmirn) [inline numba/tests/test_unicode.py]: Please consider CPython test parameters, especially symbol `\u1FFc`:
https://github.com/python/cpython/blob/201c8f79450628241574fba940e08107178dc3a5/Lib/test/test_unicode.py#L604-L605 
This symbol is not covered by this implementation.

**Comment by shssf** [comment]
> @stuartarchibald , Regarding your comment [#4442 (comment)](https://github.com/numba/numba/pull/4442#discussion_r314353276) , I do not think we can get good profit from generation of those isFoo algorithms: They are small enough and have only loop on string that can be extracted. Think that it would be easier and clear to implement them directly.
> @shssf , What do you think?

I looked comments in both PRs and didn't catch the question and idea about "isFoo"

**Comment by stuartarchibald** [comment]
> > @stuartarchibald , Regarding your comment [#4442 (comment)](https://github.com/numba/numba/pull/4442#discussion_r314353276) , I do not think we can get good profit from generation of those isFoo algorithms: They are small enough and have only loop on string that can be extracted. Think that it would be easier and clear to implement them directly.
> > @shssf , What do you think?
> 
> I looked comments in both PRs and didn't catch the question and idea about "isFoo"

It seems like all the `isFOO` methods follow approximately the same pattern, in pseudo-python:
```python
@overload_method(types.UnicodeType, 'isFOO')
def unicode_isFOO(data):

    def impl(data):
        length = len(data)
        if length == 1:
            return _PyUnicode_Is<FOO>(_get_code_point(data, 0))
        if length == 0:
            return False

        cased = False
        for idx in range(length):
            code_point = _get_code_point(data, idx)
            <FOO specific function>
    return impl
```
I think it is therefore possible to just write a generator function to simply create pretty much all of these in one go, with specializations for the functions on each branch. This is a fairly common pattern in the Numba code base. I imagine something like:
```
overload_method(types.UnicodeType, 'isFOO')(generate_isFOO(_PyUnicode_Is<FOO>, <FOO specific function>))
```
the functions can be set to be inlined at IR level or LLVM will likely just inline them anyway.

One strategy to do this might be to:
1. Implement the rest of the `_PyUnicode_IsFOO` functions in `numba.unicode_support`
2. Write the specialised functions to go in the loop section for each `IsFOO`.
3. Write the generator.
4. Wire it up.

**Comment by shssf** [comment]
> write a generator function to simply create pretty much all of these

Do you mean generation of python sources at compile time or generate string with python function and use `exec` later?
Second approach is hard to debug.

**Comment by stuartarchibald** [comment]
> > write a generator function to simply create pretty much all of these
> 
> Do you mean generation of python sources at compile time or generate string with python function and use `exec` later?
> Second approach is hard to debug.

The first, define the common template as an inner function and close over it with the specialisations, return the inner function and jit it.

File: numba/unicode.py [hunk 2]
```diff
@@ -1286,6 +1286,31 @@ def impl(a):
     return impl
 
 
+@overload_method(types.UnicodeType, 'istitle')
+def unicode_istitle(s):
```
[the remark this answers is not in the document]
Comment (Vyacheslav-Smirnov) [reply numba/unicode.py]: I do not think so due to "s" is the unicode string for which the function is called thus it is a unicode type (in other word if "s" is not a unicode this function would not be called for it)

File: numba/tests/test_unicode.py (hunk 1 above)
Comment (Vyacheslav-Smirnov) [reply numba/tests/test_unicode.py]: Will add this, thanks!

File: numba/unicode.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/unicode.py]: @Vyacheslav-Smirnov's interpretation is correct. `s` is effectively a reference to `self`, i.e. the type against which the overloaded method is bound.

**Comment by Vyacheslav-Smirnov** [comment]
Hi @stuartarchibald ,
I suppose that not all is... functions can follow to this template https://github.com/numba/numba/pull/4645#issuecomment-537903059
For example, in this case - istitle - we need 2 parameters outside the specific function thus the template should be changed already for this case. Not sure that the rest functions does not have such algorithm limitation.
Maybe implement them all and then think about how to unite them using one template if it is possible

**Comment by stuartarchibald** [comment]
> Hi @stuartarchibald ,
> I suppose that not all is... functions can follow to this template [#4645 (comment)](https://github.com/numba/numba/pull/4645#issuecomment-537903059)
> For example, in this case - istitle - we need 2 parameters outside the specific function thus the template should be changed already for this case. Not sure that the rest functions does not have such algorithm limitation.
> Maybe implement them all and then think about how to unite them using one template if it is possible

You can just wrap the parameters in a lambda/have dead code as it doesn't cost anything. If it makes it easier to do this by first writing all these out by hand and then collapsing them into a template then that's fine, my intention was to try and save everyone time/effort by avoiding that :) As the implementers, it's up to you!

File: numba/unicode.py [hunk 3]
```diff
@@ -1286,6 +1286,31 @@ def impl(a):
...
+    """
+    Implements UnicodeType.istitle()
+    """
+
+    def impl(s):
+        str_has_abc = False
```
Comment (stuartarchibald) [inline numba/unicode.py]: What's `_abc` mean here? also this alg looks a lot like something from CPython, is a citation with SHA missing?

Comment (Vyacheslav-Smirnov) [reply numba/unicode.py]: Hi,
By "str_has_abc" I have tried to say that string has alphabetical characters.
The "cased" from CPython algorithm does not express its meaning for me.
Ok, I will add the link to CPython here

**Comment by Vyacheslav-Smirnov** [comment]
Hi @stuartarchibald ,
I have applied your comment. Could you please check if all is fine and nothing else should be changed?

**Comment by stuartarchibald** [comment]
Thanks, the final patch addresses all comments, I'll rerun CI. It should be noted however that https://github.com/numba/numba/pull/4501 is likely to be merged soon and as a result any lines >80 chars will violate the `flake8` specification and consequently break the build, it might be worth addressing this now? For comments containing citations/references to github URLs, append `    # noqa: E501` to keep flake8 happy!

File: numba/unicode.py (hunk 3 above)
Comment (stuartarchibald) [reply numba/unicode.py]: I think the `cased`/`previous_is_cased` is an abbreviated form of `is_title_cased`/`previous_is_title_cased`, this algorithm is essentially looking for certain patterns in a string that may occur between title-cased and uncased/lower-case characters.

**Comment by Vyacheslav-Smirnov** [comment]
Hi @stuartarchibald ,
Add the # noqa: E501 comment to the url to CPython. Looks like all should be good now.

File: numba/unicode.py (hunk 3 above)
Comment (Vyacheslav-Smirnov) [reply numba/unicode.py]: Ok, I understand.
Then I revert names back.

Comment (stuartarchibald) [reply numba/unicode.py]: Great, thanks.

**Comment by Vyacheslav-Smirnov** [comment]
Hi,
Looks like the CI is passed. So, if nothing else should be changed, could you please approve the PR?

**Comment by Vyacheslav-Smirnov** [comment]
Hi @stuartarchibald ,
Looks like the CI is passed. So, if nothing else should be changed, could you please approve the PR?

**Comment by stuartarchibald** [comment]
Going to re-run CI now since https://github.com/numba/numba/pull/4501 got merged.

**Comment by Vyacheslav-Smirnov** [comment]
Travis build fails during install of miniconda, Windows is cancelled. Other builds are passed.

**Comment by Vyacheslav-Smirnov** [comment]
Thank you!

**Pull request merged**